### PR TITLE
mcs, tso: replace expected-primary keepalive with transient marker

### LINKED ISSUE #10875: tso: campaign may fail when expected primary key is deleted but lease persists
### What did you do?
In the TSO primary election loop, before campaigning `GetExpectedPrimaryFlag` reads the
expected primary key from etcd. When the key does not exist, it returns an empty string,
allowing the campaign to proceed unconditionally.

### What did you expect to see?
When the expected primary key is missing, the system should still check whether the
associated lease is alive before allowing the campaign to proceed, so that the old
primary's leader key does not block the new campaign.

### What did you see instead?
The expected primary key is bound to an etcd lease. In a corner case, the key may be
deleted while the associated lease remains alive. When this happens:

`GetExpectedPrimaryFlag` returns `""` → the check `len(expectedPrimary) > 0` is
bypassed → campaign proceeds → `Campaign` fails because `CreateRevision(leaderKey)`
is non-zero (old primary's key still exists) → `ErrEtcdTxnConflict` disrupts election.

### Affected Code
- `pkg/mcs/utils/expected_primary.go:39-48` — `GetExpectedPrimaryFlag` returns "" when
  key not found, without checking lease state
- `pkg/tso/allocator.go:248-261` — campaign bypassed when expectedPrimary is empty
- `pkg/mcs/scheduling/server/server.go:259-261` — same pattern
- `pkg/mcs/resourcemanager/server/server.go:174-186` — same pattern

### What version of PD are you using (`pd-server -V`)?
master

### PR DESCRIPTION
**Pull request by bufferflies**
### What problem does this PR solve?

Issue Number: Close #10875

The microservice primary used to keep an `expected_primary` etcd key alive
with a dedicated lease for its whole tenure and run a watch goroutine
(`KeepExpectedPrimaryAlive`) to learn about a transfer. That second lease
could desync from the leader key: when the key was deleted while its lease
persisted, the next campaign read an empty flag, bypassed the guard, and
then failed with `ErrEtcdTxnConflict` because the old primary's leader key
still existed — disrupting the election (#10875).

### What is changed and how does it work?

Replace the persistent keepalive + watch with a transient, self-cleaning flow:

```commit-message
- `expected_primary` is only written by the `{service}/primary/transfer`
  API now, with a TTL of a few leader leases (`TransferPrimaryLeaseMultiplier`
  times the leader lease); it is no longer kept alive for the primary's whole
  tenure.
- `TransferPrimary` marks the target first, then resigns the current primary
  by revoking its leader lease. `IsServing()` flips to false immediately, so
  the existing primary ticker steps the old primary down — the watch
  goroutine is gone.
- The campaign is made atomic via `CampaignWithCmps`: when a marker is
  observed, `expected_primary == observed-value` is appended to the
  leader-key txn, so the target wins only if the marker still names it.
- The affinity guard matches the marker against the participant's listen URLs
  exactly (`Participant.IsExpectedPrimary`) instead of a substring search of
  the serialized member value, so one member's address being a substring of
  another's value cannot misroute a transfer.
- The new primary CAS-deletes the marker once it wins and revokes the
  marker's lease, leaving neither the key nor a dangling lease (avoiding the
  "key deleted but lease persists" state from #10875).
- `Lease.GetTimeoutSeconds` falls back to `constant.DefaultLease` for a
  nil/ungranted lease, and `Lease.Close` is idempotent so resigning from both
  the transfer path and the election loop does not double-revoke or log
  spurious errors.
- Remove the now-dead `primaryWatch` / `IsPrimary` / `SetPrimaryWatch` and
  the PUT branch in `Leadership.Watch`, plus the `expectedPrimaryLease` field
  and accessors on `Participant` and the TSO `Allocator`.
```

Behavior note: if a transfer target dies before winning the campaign, the
cluster has no primary until the flag TTL (a few leader leases) expires and a
free election takes over (covered by the updated server-down test).

#### Upgrade compatibility

This is not a hard backward-incompatible change; a mixed old/new cluster
stays available and transfers still converge:

- A transfer is always executed by the node that is currently primary, using
  its own binary, so the "mark + step-down" path is never split across
  versions.
- The `expected_primary` value is the member's advertise address; both the old
  substring guard and the new exact guard accept it, so the formats interoperate.
- No double primary (both versions campaign with `CreateRevision(leaderKey)
  == 0`) and no permanent leaderless state (every marker has a finite TTL).
- An empty marker (new steady state) and a lingering kept-alive marker (old
  steady state) are both handled correctly by the other version.

The only things that require the whole cluster to be on the new version are
the strict "only the named target wins under racing transfers" guarantee and
the leader-lease-proportional fallback TTL. Recommendation: avoid invoking
`primary/transfer` during the rolling upgrade window; use it once the upgrade
completes.

### Check List

Tests

- Unit test
- Integration test
[transfer-primary-test.md](https://github.com/user-attachments/files/29686224/transfer-primary-test.md)

### Release note

```release-note
Fix a corner case where a microservice (TSO/scheduling/resource_manager)
primary election could be disrupted by an etcd transaction conflict when the
expected primary key was deleted while its lease still persisted.
```

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Primary transfers now use a transient “expected primary” affinity window to conditionally allow campaigning for smoother leadership handoffs.
- **Bug Fixes**
  - Removed stale expected-primary keepalive/watch behavior and clear the affinity guard after a successful campaign.
  - Leadership step-down now relies on actual serving/lease state rather than a separate auxiliary signal.
  - Updated expected-primary transfer/eligibility handling, including the TTL/timeout calculation behavior.
- **Tests**
  - Updated primary-transfer integration assertions and recovery timing for the new transient affinity behavior.
  - Adjusted real-cluster etcd key expectations to reflect expected-primary being transient only.

### CODE REVIEW

File: pkg/mcs/utils/expected_primary.go [hunk 1]
```diff
@@ -64,92 +61,93 @@ func markExpectedPrimaryFlag(client *clientv3.Client, msParam *keypath.MsParam,
...
-	case exitPrimary <- struct{}{}:
+	}
+	log.Info("delete expected primary flag", zap.String("primary-path", path))
+	// Revoke the lease the flag was bound to, if any, so no lease is leaked.
+	kvs := resp.Responses[0].GetResponseRange().GetKvs()
+	if len(kvs) == 0 || kvs[0].Lease == 0 {
 		return
 	}
+	leaseID := clientv3.LeaseID(kvs[0].Lease)
+	if _, err := client.Revoke(client.Ctx(), leaseID); err != nil {
```
Comment (rleungx) [inline pkg/mcs/utils/expected_primary.go]: This revoke runs after the new primary has won but before it is promoted to serving. Please use a timeout context here; otherwise a hung revoke RPC can block serving even though this cleanup is meant to be best-effort.

File: pkg/mcs/utils/expected_primary_test.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	"github.com/tikv/pd/pkg/utils/keypath"
+)
+
+// TestDeleteExpectedPrimaryFlagRevokesLease reconstructs the corner case from issue
+// #10875. The expected primary flag is bound to an etcd lease; if the key were
+// deleted while its lease lingered, a later campaign would read an empty flag, skip
+// the affinity guard, and then fail with ErrEtcdTxnConflict against the still-present
+// leader key. DeleteExpectedPrimaryFlag must therefore remove both the key and its
+// lease, so the "key deleted but lease persists" state can never exist.
+func TestDeleteExpectedPrimaryFlagRevokesLease(t *testing.T) {
```
Comment (rleungx) [inline pkg/mcs/utils/expected_primary_test.go:41]: CI is failing here because this new test file is not covered by goleak. Please add a TestMain with goleak.VerifyTestMain, using testutil.LeakOptions if needed for the etcd test cluster.

File: tests/integrations/mcs/members/member_test.go [hunk 3]
```diff
@@ -460,11 +460,25 @@ func (suite *memberTestSuite) TestTransferPrimaryWhileLeaseExpiredAndServerDown(
...
-		// Primary should be different with before
-		onlyPrimary, err := suite.pdClient.GetMicroservicePrimary(suite.ctx, service)
-		re.NoError(err)
-		re.NotEqual(newPrimary, onlyPrimary)
+		// The transfer picked a random target (new_primary is empty). If that target
+		// happens to be the node we just closed, the expected primary flag keeps the
+		// other replicas backing off until it expires, so recovery can take up to the
+		// flag's TTL (a few leader leases) before a free re-election elects a live
+		// primary. Wait long enough to cover that worst case.
+		recoverWait := time.Duration(mcs.TransferPrimaryLeaseMultiplier*mcs.DefaultLease)*time.Second + 10*time.Second
```
Comment (rleungx) [inline tests/integrations/mcs/members/member_test.go:470]: Could we also cover the non-default keyspace group recovery path here? The happy path is covered, but this marker-expiry/free-election case currently only runs against the default TSO group.

**Review by rleungx — APPROVED** [review]
The rest LGTM

File: pkg/election/lease.go [hunk 4]
```diff
@@ -25,6 +25,7 @@ import (
 	"github.com/pingcap/log"
 
 	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/pkg/mcs/utils/constant"
```
Comment (lhy1024) [inline pkg/election/lease.go]: pkg/election is a generic lower-level election package, so importing pkg/mcs/utils/constant here for the GetTimeoutSeconds fallback makes the layering unclear and increases the risk of future import cycles from mcs -> election. Please keep the fallback constant in the election package itself, or let the MCS caller handle the nil/ungranted lease default.

File: tests/integrations/mcs/members/member_test.go [hunk 5]
```diff
@@ -460,11 +460,25 @@ func (suite *memberTestSuite) TestTransferPrimaryWhileLeaseExpiredAndServerDown(
 		nodes[newPrimary].Close()
 		re.NoError(failpoint.Disable("github.com/tikv/pd/pkg/election/skipGrantLeader"))
 
-		tests.WaitForPrimaryServing(re, nodes)
-		// Primary should be different with before
-		onlyPrimary, err := suite.pdClient.GetMicroservicePrimary(suite.ctx, service)
-		re.NoError(err)
-		re.NotEqual(newPrimary, onlyPrimary)
+		// The transfer picked a random target (new_primary is empty). If that target
```
Comment (lhy1024) [inline tests/integrations/mcs/members/member_test.go]: This test intends to cover “the transfer target dies before winning, then other replicas recover after the marker TTL expires”, but the request still sends new_primary: "". The actual target is randomly selected inside TransferPrimary, so the node closed here is not necessarily the one recorded in the expected_primary marker. The test can pass after closing an unrelated follower. Please either set new_primary to newPrimary, or read expected_primary after the transfer and close the actual marker target.

**Comment by lhy1024** [comment]
In `pkg/mcs/utils/expected_primary.go`, `GetExpectedPrimaryFlag` should not collapse read failures into an empty marker.

Collapsing a read error into an empty expected-primary marker weakens the core guarantee of this PR. Callers treat `""` as “no marker” and campaign without `ExpectedPrimaryCmp`; if the marker actually exists but this read fails, a non-target member can still win the election. Please make `GetExpectedPrimaryFlag` return `(string, error)` and have the election loop back off/retry on read errors instead of treating them as an absent marker.

**Comment by lhy1024** [comment]
One more stale-comment issue outside the changed diff: please also update the `expected_primary` comment in `pkg/utils/keypath/absolute_key_path.go`. It still says a successful primary campaign sets `expected_primary` and transfer revokes the previous lease, while this PR changes the semantics to “only the transfer API writes a transient marker, and the winner deletes it after campaigning”. The stale comment can mislead future maintainers.

**Review by lhy1024 — APPROVED** [review]

**Comment by bufferflies** [comment]
hold，wait detail test for it
